modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionScriptEngine.java

```diff
@@ -362,4 +299,106 @@ private ScriptException convertToScriptException(String message, String source,
+                    dateAccessor = true;
+                } else if (parts[3].type == VariableContext.Type.MEMBER) {
+                    variablename = parts[3].text;
+                    dateAccessor = true;
+                }
+            }
+            if (!dateAccessor) {
+                throw new IllegalArgumentException(
+                    "Variable [" + variable + "] does not follow an allowed format of either doc['field'] or doc['field'].method()"
+                );
+            }
+        }
+
+        String fieldname = parts[1].text;
+        MappedFieldType fieldType = lookup.doc().mapperService().fullName(fieldname);
+
+        if (fieldType == null) {
+            throw new ParseException("Field [" + fieldname + "] does not exist in mappings", 5);
+        }
+
+        IndexFieldData<?> fieldData = lookup.doc().getForField(fieldType);
+        final ValueSource valueSource;
+        if (fieldType instanceof GeoPointFieldType) {
+            // geo
+            if (methodname == null) {
+                valueSource = GeoField.getVariable(fieldData, fieldname, variablename);
+            } else {
+                valueSource = GeoField.getMethod(fieldData, fieldname, methodname);
+            }
+        } else if (fieldType instanceof DateFieldMapper.DateFieldType) {
+            if (dateAccessor) {
+                // date object
+                if (methodname == null) {
+                    valueSource = DateObject.getVariable(fieldData, fieldname, variablename);
+                } else {
+                    valueSource = DateObject.getMethod(fieldData, fieldname, methodname);
+                }
+            } else {
+                // date field itself
+                if (methodname == null) {
+                    valueSource = DateField.getVariable(fieldData, fieldname, variablename);
+                } else {
+                    valueSource = DateField.getMethod(fieldData, fieldname, methodname);
+                }
+            }
+        } else if (fieldData instanceof IndexNumericFieldData) {
+            // number
+            if (methodname == null) {
+                valueSource = NumericField.getVariable(fieldData, fieldname, variablename);
+            } else {
+                valueSource = NumericField.getMethod(fieldData, fieldname, methodname);
+            }
+        } else {
+            throw new ParseException("Field [" + fieldname + "] must be numeric, date, or geopoint", 5);
+        }
+        return valueSource;
+    }
+
+    // TODO: document and/or error if vars contains _score?
+    // NOTE: by checking for the variable in vars first, it allows masking document fields with a global constant,
+    // but if we were to reverse it, we could provide a way to supply dynamic defaults for documents missing the field?
+    private static void bindFromVars(@Nullable final Map<String, Object> vars, final SimpleBindings bindings, final String variable) throws ParseException {
```

Comment:
I know "vars" is the terminology used in the code originally, but irrc this is actually `params`? I think that naming would be better.